integration-tests/tests/src/tests/objects.ts

```diff
@@ -17,244 +17,1323 @@
       });
 
-      expect(john._id.equals(_id)).equals(true);
-      expect(john.name).equals("Mr. John Doe");
-      expect(john.age).equals(43);
+      // can only set property in write transaction
+      expect(function () {
+        obj.boolCol = false;
+      }).throws;
 
-      expect(() =>
-        realm.write(() => {
-          realm.create<IPersonWithId>(
-            PersonSchemaWithId.name,
-            { _id, name: "John Doe", age: 42 },
-            Realm.UpdateMode.Never,
-          );
-        }),
-      ).throws(
-        `Attempting to create an object of type '${PersonSchemaWithId.name}' with an existing primary key value '${_id}'.`,
-      );
+      expect(obj.boolCol).equals(true, "bool value changed outside transaction");
+
+      this.realm.write(function () {
+        //@ts-expect-error assign string to non matching type.
+        expect(() => (obj.boolCol = "cat")).throws;
+        //@ts-expect-error assign string to non matching type.
+        expect(() => (obj.intCol = "dog")).throws;
+
+        // Non-optional properties should complain about null
+        for (const name of ["boolCol", "intCol", "floatCol", "doubleCol", "stringCol", "dataCol", "dateCol"]) {
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          expect(() => (obj[name] = null), `Setting ${name} to null should throw`).throws;
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          expect(() => (obj[name] = undefined), `Setting ${name} to undefined should throw`).throws;
+        }
+
+        // Optional properties should allow it
+        for (const name of [
+          "optBoolCol",
+          "optIntCol",
+          "optFloatCol",
+          "optDoubleCol",
+          "optStringCol",
+          "optDataCol",
+          "optDateCol",
+          "objectCol",
+        ]) {
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          obj[name] = null;
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          expect(obj[name]).equals(null);
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          obj[name] = undefined;
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          expect(obj[name]).equals(null);
+        }
 
-      // Expect only one instance of 'PersonSchemaWithId' in db after all updates
-      const persons = realm.objects(PersonSchemaWithId.name);
-      expect(persons.length).equals(1);
+        function tryAssign(name: string, value: any) {
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
+          const prop = AllTypesSchema.properties[name];
+          const type = typeof prop == "object" ? prop.type : prop;
+          //@ts-expect-error TYPEBUG: indexing with string is not allowed by typesystem.
```

Comment:
If we used a class-based model it might not complain

Reply:
I'll leave this as is for now (this error exists in multiple test files already and it could be worth doing them altogether in a single commit). If you guys haven't discussed it already I think it'd be worth considering wether we should support this syntax in TS also or if we should encourage consumers to use the class based model instead.